src/commands/classified/pipeline.rs

```diff
@@ -1,107 +1,74 @@
+use nu_source::Text;
 use std::sync::atomic::Ordering;
 
-#[derive(Debug, Clone)]
-pub(crate) struct Pipeline {
-    pub(crate) commands: ClassifiedCommands,
-}
-
-impl Pipeline {
-    pub fn commands(list: Vec<ClassifiedCommand>, span: impl Into<Span>) -> Pipeline {
-        Pipeline {
-            commands: ClassifiedCommands {
-                list,
-                span: span.into(),
-            },
-        }
-    }
-
-    pub(crate) async fn run(
-        self,
-        ctx: &mut Context,
-        mut input: ClassifiedInputStream,
-        line: &str,
-    ) -> Result<(), ShellError> {
-        let mut iter = self.commands.list.into_iter().peekable();
-
-        loop {
-            let item: Option<ClassifiedCommand> = iter.next();
-            let next: Option<&ClassifiedCommand> = iter.peek();
-
-            input = match (item, next) {
-                (Some(ClassifiedCommand::Dynamic(_)), _)
-                | (_, Some(ClassifiedCommand::Dynamic(_))) => {
-                    return Err(ShellError::unimplemented("Dynamic commands"))
-                }
+pub(crate) async fn run_pipeline(
+    pipeline: ClassifiedPipeline,
+    ctx: &mut Context,
+    mut input: ClassifiedInputStream,
+    line: &str,
+) -> Result<(), ShellError> {
+    let mut iter = pipeline.commands.list.into_iter().peekable();
+
+    loop {
+        let item: Option<ClassifiedCommand> = iter.next();
+        let next: Option<&ClassifiedCommand> = iter.peek();
+
+        input = match (item, next) {
+            (Some(ClassifiedCommand::Dynamic(_)), _) | (_, Some(ClassifiedCommand::Dynamic(_))) => {
+                return Err(ShellError::unimplemented("Dynamic commands"))
+            }
 
-                (Some(ClassifiedCommand::Expr(_)), _) | (_, Some(ClassifiedCommand::Expr(_))) => {
-                    return Err(ShellError::unimplemented("Expression-only commands"))
-                }
+            (Some(ClassifiedCommand::Expr(_)), _) | (_, Some(ClassifiedCommand::Expr(_))) => {
+                return Err(ShellError::unimplemented("Expression-only commands"))
+            }
 
-                (Some(ClassifiedCommand::Internal(left)), _) => {
-                    let stream = left.run(ctx, input, Text::from(line))?;
-                    ClassifiedInputStream::from_input_stream(stream)
-                }
+            (Some(ClassifiedCommand::Internal(left)), _) => {
+                let stream = run_internal_command(left, ctx, input, Text::from(line)).await?;
+                ClassifiedInputStream::from_input_stream(stream)
+            }
 
-                (Some(ClassifiedCommand::External(left)), Some(ClassifiedCommand::External(_))) => {
-                    left.run(ctx, input, StreamNext::External).await?
-                }
+            (Some(ClassifiedCommand::External(left)), Some(ClassifiedCommand::External(_))) => {
+                run_external_command(left, ctx, input, StreamNext::External).await?
```

Comment:
I'm a little confused about this bit. I think the intent of the PR this came from was to begin to consolidate the command types more. This seems to segregate with different helper commands.

Reply:
the `.run` methods were previously inherent methods on the two command types, but the command types were moved into the parser. In general, when there were only a handful of uses of an inherent method, I addressed these issues for now by moving the inherent methods to functions.